src/utils.py

```diff
@@ -8,86 +8,116 @@
+    CERTIFICATE_TRANSFER_INTEGRATION_NAME,
+    HYDRA_ENDPOINTS_INTEGRATION_NAME,
+    INGRESS_INTEGRATION_NAME,
+    KRATOS_INFO_INTEGRATION_NAME,
+    OPENFGA_INTEGRATION_NAME,
+    PEER_INTEGRATION_NAME,
+    WORKLOAD_CONTAINER,
+)
+
 logger = logging.getLogger(__name__)
 
 CharmEventHandler = TypeVar("CharmEventHandler", bound=Callable[..., Any])
-ConditionEvaluation = tuple[bool, str]
-Condition = Callable[[CharmBase], ConditionEvaluation]
+Condition = Callable[[CharmBase], bool]
 
 
-def container_not_connected(charm: CharmBase) -> ConditionEvaluation:
-    """Condition to validate workload container connectivity."""
-    not_connected = not charm._container.can_connect()
-    return not_connected, ("Container is not connected yet" if not_connected else "")
+def leader_unit(func: CharmEventHandler) -> CharmEventHandler:
+    """A decorator, applied to any event hook handler, to validate juju unit leadership."""
 
+    @wraps(func)
+    def wrapper(charm: CharmBase, *args: Any, **kwargs: Any) -> Optional[Any]:
+        if not charm.unit.is_leader():
+            return None
 
-def integration_not_exists(integration_name: str) -> Condition:
-    """A factory of conditions of integration existence validation."""
+        return func(charm, *args, **kwargs)
 
-    def wrapped(charm: CharmBase) -> ConditionEvaluation:
-        not_exists = not charm.model.relations[integration_name]
-        return not_exists, (f"Missing integration {integration_name}" if not_exists else "")
+    return wrapper  # type: ignore[return-value]
 
-    return wrapped
 
+def integration_existence(integration_name: str) -> Condition:
+    """A factory of integration existence condition."""
 
-def block_when(*conditions: Condition) -> Callable[[CharmEventHandler], CharmEventHandler]:
-    """A factory of decorators that block juju unit when a certain condition validation fails."""
+    def wrapped(charm: CharmBase) -> bool:
+        return bool(charm.model.relations[integration_name])
 
-    def decorator(func: CharmEventHandler) -> CharmEventHandler:
-        """A decorator, applied to any event hook handler, to block juju unit."""
+    return wrapped
 
-        @wraps(func)
-        def wrapper(charm: CharmBase, *args: EventBase, **kwargs: Any) -> Optional[Any]:
-            event, *_ = args
-            logger.debug(f"Handling event: {event}.")
 
-            for condition in conditions:
-                not_met, msg = condition(charm)
-                if not_met:
-                    charm.unit.status = BlockedStatus(msg)
-                    return None
+peer_integration_existence = integration_existence(PEER_INTEGRATION_NAME)
+kratos_integration_existence = integration_existence(KRATOS_INFO_INTEGRATION_NAME)
+hydra_integration_existence = integration_existence(HYDRA_ENDPOINTS_INTEGRATION_NAME)
+openfga_integration_existence = integration_existence(OPENFGA_INTEGRATION_NAME)
+ingress_integration_existence = integration_existence(INGRESS_INTEGRATION_NAME)
```

Comment:
I would rename these to
```suggestion
peer_integration_exists = integration_existence(PEER_INTEGRATION_NAME)
kratos_integration_exists = integration_existence(KRATOS_INFO_INTEGRATION_NAME)
hydra_integration_exists = integration_existence(HYDRA_ENDPOINTS_INTEGRATION_NAME)
openfga_integration_exists = integration_existence(OPENFGA_INTEGRATION_NAME)
ingress_integration_exists = integration_existence(INGRESS_INTEGRATION_NAME)
```